docs/sw.1.scd
sw(1)

# NAME

sw - terminal stopwatch

# SYNOPSIS

*sw* [_OPTIONS_] [_NAME_]

# DESCRIPTION

*sw* is a terminal stopwatch that runs as a shell.

# PARAMETERS

_NAME_
	Set the stopwatch name.

	This is equivalent to using the _Name_ command.

*-v*, *--no-visual-cues*
	Disable text-based graphics and visual cues.

	To toggle this value at runtime, use the _Visuals_ command.

	For more details, see the section on visual cues.

*-c*, *--no-colors*
	Disable the use of colors in output.

	For more details, see the section on colors.

*--no-tty-check*
	Disable the TTY check.

	Typically, *sw* will check if standard output and input are terminals,
	and exit abnormally if either aren't. This flag disables the check.

*-V*, *--version*
	Display the version of *sw*.

*--help*
	Display usage information.

# COMMANDS

The shell interface allows you to manipulate a stopwatch by issuing
commands and responding to prompts that may follow.

The following table is a quick overview of the available commands. The "input"
field shows which key combination maps to the command. The command's name is
also a valid input.

[[ Command name
:[ Input
:[ Description
|[ Help
:  h
:  Show help
|  Display
:  <Enter>
:  Display elapsed time
|  Toggle
:  s
:  Toggle stopwatch
|  Reset
:  r
:  Reset stopwatch
|  Change
:  c
:  Change elapsed time
|  Offset
:  o
:  Offset elapsed time
|  Name
:  n
:  Set stopwatch name
|  Precision
:  p
:  Set display precision
|  Visuals
:  v
:  Toggle visual cues
|  License
:  l
:  Print license info
|  Quit
:  q
:  Abandon all Data

## HELP

Lists the available commands and their description.

Each line of the output corresponds to a command, and contains the required key
combination and a description of the command.

## DISPLAY

Displays the state of the stopwatch.

The first line of output contains the elapsed time. The subsecond precision of
this output is affected by the _Precision_ command.

The second line indicates whether the stopwatch is running or stopped. The
colors used to indicate this are an exception to their typical meaning. For more
information, see the the section on colors.

## TOGGLE

Toggles whether the stopwatch is running or stopped.

If the stopwatch is running, it is stopped. If the new elapsed time overflows,
it prints a warning.

If stopped, it starts the stopwatch. Then it prints the duration that has
elapsed since the last time it was stopped. This may be useful if you forget to
start the stopwatch, since you can start it and it'll tell you how long the
stopwatch was idling. The precision of this output is affected by the
_Precision_ command.

## RESET

Resets the stopwatch.

This command stops the stopwatch and sets its elapsed time to zero.

## CHANGE

Stops and sets the elapsed time.

This command prompts you for a new elapsed time.

The duration will be parsed according to the duration formats. It must be
positive. If the duration is invalid, it will be printed back with the errors
colored red.

## OFFSET

Offsets the elapsed time by a positive or negative duration.

This command prompts you for a positive or negative duration. If it parses
according to the duration formats, the elapsed time will be offset by it without
changing whether the stopwatch is running. Otherwise, the invalid duration will
be printed with errors indicated in red.

## NAME

Sets the name of the stopwatch.

This command will prompt you to name the stopwatch.

The name is purely for display: it's literal text that prefixes the prompt. It
may be useful for distinguishing the meaning of the stopwatch if you have
multiple instances open.

## PRECISION

Sets the decimal precision used to display the subsecond component of durations.

Several commands output formatted elapsed times and durations. This command is
used to configure the decimal precision of this output.

The default precision is 2, meaning two subsecond digits will be displayed.

## VISUALS

Toggles whether text-based graphics and visual cues are enabled (they are by
default).

This parameter can be controlled with the *--no-visual-cues* flag.

For more details, see the section on visual cues.

## LICENSE

Prints license and copyright information for *sw* and its direct dependencies.

## QUIT

Terminates the program with a successful exit code, abandoning all stopwatch
state.

# PROMPT

*sw* displays a prompt before commands are entered.

The prompt contains the name of the stopwatch (see the _Name_ command). If
visual cues are enabled, the name is followed by a character indicating the
stopwatch state. This indicator is ';' if the stopwatch is stopped and '\*' if
it's running.

For example, if the stopwatch is named "work" and stopped, it will prompt with
the following with visual cues enabled:

```
work ; 
```

And if it's running:

```
work * 
```

With the same stopwatch but visual cues disabled, it will prompt with this:

```
work. 
```

# DURATION INPUT FORMAT

The _Offset_ and _Change_ commands accept a duration as input. The following
formats are supported.

## SHORT FORMAT

```
number unit
```

_number_ is a whole or decimal number, and _unit_ is one of "s", "m", or "h",
meaning seconds, minutes, and hours respectively.

Whitespace is ignored, so "1s" is just as valid as "1 s".

## LONG FORMAT

```
sign hours : minutes : seconds . subseconds
```

The details shouldn't be surprising, it's a superset of how durations are
displayed.

_hours_, _minutes_, _seconds_ and _subseconds_ are all integers.

_sign_ is optional, and one of "+" or "-". It indicates whether the duration is
positive or negative, defaulting to positive.

It's okay to omit separators and values. Rightmost values are the most
important, so the meaning of the input will be inferred from right to left.

Some examples of terse inputs:

- ":5" and "::5" represent 5 seconds
- ":5:" represents 5 minutes
- ":.6" represents 0.6 seconds
- "1::1.1" represents 1 hour and 1.1 seconds

It's also okay to add whitespace between separators.

# DURATION OUTPUT FORMAT

With visual cues enabled, durations are formatted as
_hours:minutes:seconds.subseconds_. Otherwise they're formatted as they would be
spoken. For example, "00:05:32.98" would be formatted as "5 minutes, 32.98
seconds". If the precision is zero, this will simply be "5 minutes, 32 seconds".

# COLORS

*sw* will use colors if they are supported by the terminal, or as directed by
the *--no-colors* flag.

The following table documents the meaning of each color as it appears:

[[ Color
:[ Meaning
|  Magenta
:  Information about program state change
|  Cyan
:  Information, but state is unchanged
|  Yellow
:  Warning
|  Red
:  Error

The indication for whether the stopwatch is running or stopped in the _Display_
command is an exception to this, with "running" colored green and "stopped"
colored yellow.

# VISUAL CUES

Visual cues and text-based graphics are enabled by default, but can be disabled
using the _Visuals_ command and the *--no-visual-cues* flag.

The following aspects of the interface are affected by this parameter:
- Prompt
- Duration parsing error messages
- Duration output format

# AUTHORS

Maintained by Ula Shipman <ula.hello@mailbox.org>.

Source code and issue tracker can be found at
<https://github.com/ulahello/sw>.


CHANGELOG.md
# Changelog

## [Unreleased]
### Changed
* renamed duration output formats:
  * stopwatch format is now long format
  * unit format is now short format

## [0.15.0] - 2025-01-30
### Changed
* `sw` now quits when EOF is received in place of a command

### Fixed
* fixed inconsistent saturation behavior between positive and negative offsets
  * now, both will saturate.

## [0.14.3] - 2024-09-23
### Fixed
* fixed documentation error: stopwatch format and unit format were interchanged

## [0.14.2] - 2024-09-12
### Fixed
* fixed 0.14.1 regression in unit format subsecond parsing

## [0.14.1] - 2024-09-12
### Changed
* more lax duration parsing
  * truncate excess subsecond digits instead of raising error
  * don't error when minutes or seconds exceed maximum
* replaced `libsw` dependency with `libsw_core`

### Removed
* removed `deny.toml`

## [0.14.0] - 2024-08-28
### Added
* added `name` positional argument to set stopwatch name
* added icon

### Changed
* reuse strings to reduce frequent allocations

### Removed
* removed nix files

## [0.13.3] - 2023-07-18
### Added
* added parse error for unexpected negative duration
* added `default.nix` and `shell.nix` for building with nix

### Changed
* replaced `is-terminal` dependency with `std::io::IsTerminal`
  * MSRV bumped to `1.70.0`

### Fixed
* replaced int overflow errors in duration parsing with duration overflow errors
* fixed hyperlink formatting in README

## [0.13.2] - 2023-06-03
### Fixed
* fixed `sw -V` requiring tty
* fixed `sw -V` writing to `stderr` instead of `stdout`

## [0.13.1] - 2023-04-05
### Added
* added tty check
* added `--no-tty-check` flag to disable tty check

### Fixed
* fixed unexpected negative offset behavior while elapsed time is overflowing
* fixed incorrect MSRV (now compiles as expected)

## [0.13.0] - 2023-04-03
### Added
* added `--no-colors` (`-c`) flag to disable color output to the terminal
* added `--version` (`-V`) flag to display version
* added suggestions for similar command names after unknown command error

### Changed
* make reset message more clearly indicate that it also stops the stopwatch
* changed fatal error prefix from "fatal: " to "fatal error: "
* added to cli description
* redirect precision parsing overflow errors to clamping warnings
* optimized output to perform less syscalls
* **BREAKING:** changed unit format from float to a integer or decimal number
* **BREAKING:** changed `--no-visual-cues` short name from `-x` to `-v`
* MSRV lowered to `1.61.0`

### Fixed
* fixed panic on fatal errors
* fixed inconsistent duration formatting
* fixed unreachable time of check bug
  * when toggling a running and overflowing stopwatch with a non-monotonic Instant, the new elapsed time could be slightly inaccurate compared to the time that overflow was checked. this is unreachable because we use `std::instant::Instant`, which is monotonic.
* only display elapsed time since last stop if it doesn't overflow

## [0.12.0] - 2023-03-20
### Added
* added license information of direct dependencies to license command output

### Changed
* improved readability of no visual cues duration format
* optimized output to emit less color resets

## [0.11.0] - 2023-03-03
### Added
* rewrote README
* added `--no-visual-cues` (`-x`) flag to remove text-based graphics and visual cues
* added `v` command to toggle visual cues
* added status messages for all commands
  * commands should never output nothing
* added help message for parsing unit format when value is missing or invalid

### Changed
* change help command format
  * no longer a markdown table, it's very plain now

## [0.10.0] - 2023-02-19
### Added
* support for HH:MM:SS.ss duration format
* overhauled error messages
* added message when subtracting clamps to zero

### Changed
* changed time display formatting
* changed prompt from "<" and ">" to ";" and "*"
* changed empty precision input to reset to default
* check for overflow when applying offset instead of saturating
* check for overflow when toggling the stopwatch
* check for overflow when displaying elapsed time
* MSRV bumped to `1.66.1`

## [0.9.2] - 2022-09-05
### Fixed
* fixed portability of colors

## [0.9.1] - 2022-09-05
### Changed
* refactored internals

### Fixed
* made "X seconds since start" message more accurate

### Removed
* removed dependency on `log` crate

## [0.9.0] - 2022-08-09
### Added
* parse duration and unit together

## [0.8.3] - 2022-08-07
### Changed
* moved stopwatch to `libsw` crate

## [0.8.2] - 2022-06-03
### Changed
* improved help message formatting

### Fixed
* improved exit code portability
  * now using `std` defined failure/success codes

### Fixed
* fixed error message formatting
  * error messages weren't followed with two newlines, like normal output

## [0.8.1] - 2022-05-02
### Changed
* refactored internals

## [0.8.0] - 2022-04-25
### Added
* added precision command
  * allows user to change the display precision on the fly
* made display command prettier
* print time since last stop

## [0.7.0] - 2022-04-24
### Added
* colored logs

## [0.6.3] - 2022-04-24
### Changed
* changed `quit` help message

### Fixed
* fixed escape control characters when taking input

## [0.6.2] - 2022-04-12
### Fixed
* fixed incorrect stopwatch subtraction behavior

## [0.6.1] - 2022-04-12
### Added
* added license command `l`
* name command: improved help and added status messages

### Changed
* buffered input so memory allocation never depends on user input

## [0.6.0] - 2022-04-11
### Added
* added stopwatch naming
* print license in splash text

## [0.5.0] - 2022-04-09
### Added
* prompt now indicates stopwatch status

### Fixed
* fixed panic on invalid float to duration conversion
* replaced calls to functions which may (but don't) panic

## [0.4.0] - 2022-04-02
### Added
* changed set command to `c` (c for change)
* changed offset command to `o`

## [0.3.0] - 2022-04-02
### Added
* added offset command: `+`
  * allows the user to add or subtract from the total elapsed time

src/parse/sw.rs
// sw: terminal stopwatch
// copyright (C) 2022-2023 Ula Shipman <ula.hello@mailbox.org>
// licensed under GPL-3.0-or-later

use unicode_segmentation::{GraphemeIndices, UnicodeSegmentation};

use core::iter::{Peekable, Rev};
use core::num::{IntErrorKind, ParseIntError};
use core::time::Duration;
use core::{fmt, ops};

use super::{
    ByteSpan, ErrKind, ParseErr, ParseFracErr, ReadDur, MIN_PER_HOUR, SEC_PER_HOUR, SEC_PER_MIN,
};

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) enum SwErrKind {
    UnexpectedColon,
    UnexpectedDot(Group),
    UnexpectedSign { is_neg: bool },
    Int { group: Group, err: ParseIntError },
    DurationOverflow(Group),
}

impl SwErrKind {
    pub(crate) fn has_help_message(&self) -> bool {
        match self {
            Self::UnexpectedColon
            | Self::UnexpectedDot(_)
            | Self::DurationOverflow(_)
            | Self::Int { .. } => true,

            Self::UnexpectedSign { .. } => false,
        }
    }
}

impl fmt::Display for SwErrKind {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        if f.alternate() {
            match self {
                Self::UnexpectedColon => {
                    write!(f, "there is no colon before {}", Group::Hours)
                }
                Self::UnexpectedDot(group) => {
                    assert_ne!(*group, Group::SecondsSub);
                    if *group == Group::SecondsInt {
                        write!(
                            f,
                            "decimal point was already given for {}",
                            Group::SecondsSub
                        )
                    } else {
                        write!(
                            f,
                            "found in {group}, but only {} can have fractional values",
                            Group::SecondsInt
                        )
                    }
                }
                Self::DurationOverflow(_) => {
                    write!(f, "this duration is too large to be represented")
                }
                Self::Int { group, err: _ } => write!(f, "{group} are parsed as an integer"),

                Self::UnexpectedSign { .. } => {
                    unreachable!()
                }
            }
        } else {
            match self {
                Self::UnexpectedColon => write!(f, "unexpected colon"),
                Self::UnexpectedDot(_) => write!(f, "unexpected decimal point"),
                Self::UnexpectedSign { is_neg: _ } => {
                    write!(f, "sign must be given at the beginning")
                }
                Self::Int { group: _, err } => write!(f, "{err}"),
                Self::DurationOverflow(group) => {
                    write!(f, "duration overflow while parsing {group}")
                }
            }
        }
    }
}

impl ReadDur {
    pub fn parse_as_sw(s: &str, allow_neg: bool) -> Result<Self, ParseErr<'_>> {
        /* split string into groups of hours, minutes, etc */
        let mut neg_span = None;
        let (groups, is_neg): (Groups, bool) = {
            // NOTE: the lexer scans IN REVERSE
            let mut lexer = SwLexer::new(s).peekable();
            let mut cur = Group::SecondsSub;
            let mut groups = Groups::new(s);
            let mut is_neg = None;
            while let Some(token) = lexer.next() {
                match (cur, token.typ) {
                    (Group::SecondsSub, SwTokenKind::Colon) => {
                        // turns out the cur has been SecondsInt this whole
                        // time!
                        let tmp = groups[cur];
                        groups[cur] = ByteSpan::new(0, 0, s);
                        cur = Group::SecondsInt;
                        groups[cur] = tmp;

                        // now that we're SecondsInt and encountering a colon,
                        // transition.
                        cur = Group::Minutes;
                    }
                    (Group::SecondsSub, SwTokenKind::Dot) => cur = Group::SecondsInt,
                    (Group::SecondsSub, SwTokenKind::Data) => {
                        // NOTE: the question im asking is: "are there are no
                        // colons after this current token?" the way im asking
                        // it is "is this the last token or is the next token a
                        // sign, regardless of whether its the last because
                        // signs are only correctly parsed as the last"
                        let next_typ = lexer.peek().map(|token| token.typ);
                        if next_typ.is_none()
                            || next_typ == Some(SwTokenKind::Pos)
                            || next_typ == Some(SwTokenKind::Neg)
                        {
                            cur = Group::SecondsInt;
                        }
                        groups[cur] = token.span;
                    }

                    (Group::SecondsInt, SwTokenKind::Colon) => cur = Group::Minutes,

                    (Group::Minutes, SwTokenKind::Colon) => cur = Group::Hours,

                    (Group::Hours, SwTokenKind::Colon) => {
                        return Err(ParseErr::new(token.span, SwErrKind::UnexpectedColon));
                    }

                    (_, SwTokenKind::Data) => groups[cur] = token.span,
                    (_, SwTokenKind::Dot) => {
                        return Err(ParseErr::new(token.span, SwErrKind::UnexpectedDot(cur)));
                    }

                    (_, SwTokenKind::Pos) => {
                        if lexer.peek().is_none() {
                            is_neg = Some(false);
                        } else {
                            return Err(ParseErr::new(
                                token.span,
                                SwErrKind::UnexpectedSign { is_neg: false },
                            ));
                        }
                    }
                    (_, SwTokenKind::Neg) => {
                        if lexer.peek().is_none() {
                            is_neg = Some(true);
                            neg_span = Some(token.span);
                        } else {
                            return Err(ParseErr::new(
                                token.span,
                                SwErrKind::UnexpectedSign { is_neg: true },
                            ));
                        }
                    }
                }
            }
            (groups, is_neg.unwrap_or(false))
        };

        if !allow_neg && is_neg {
            return Err(ParseErr::new(neg_span.unwrap(), ErrKind::Negative));
        }

        /* parse group substrings into an actual duration */
        let mut dur = Duration::ZERO;

        // hours, minutes, seconds (whole)
        for (group, sec_per_unit) in [
            (Group::Hours, u64::from(SEC_PER_HOUR)),
            (Group::Minutes, u64::from(SEC_PER_MIN)),
            (Group::SecondsInt, 1),
        ] {
            let span = groups[group];
            let to_parse = span.get().trim();
            /* NOTE: we're trimming after we get the span, meaning the to_parse
             * doesn't reflect the span. */
            if !to_parse.is_empty() {
                match to_parse.parse::<u64>() {
                    Ok(units) => {
                        let secs = units.checked_mul(sec_per_unit).ok_or_else(|| {
                            ParseErr::new(span, SwErrKind::DurationOverflow(group))
                        })?;
                        dur = dur.checked_add(Duration::from_secs(secs)).ok_or_else(|| {
                            ParseErr::new(span, SwErrKind::DurationOverflow(group))
                        })?;
                    }

                    Err(err) => return Err(ParseErr::new(span, SwErrKind::Int { group, err })),
                }
            }
        }

        // subseconds
        {
            let group = Group::SecondsSub;
            let span = groups[group];
            let to_parse = span.get();
            if !to_parse.trim().is_empty() {
                let nanos =
                    super::parse_frac(to_parse, crate::MAX_NANOS_CHARS).map_err(|frac_err| {
                        match frac_err {
                            ParseFracErr::ParseDigit { idx, len, err } => {
                                let mut span = span;
                                span.shift_start_right(idx);
                                span.len = len;
                                assert_ne!(*err.kind(), IntErrorKind::PosOverflow);
                                ParseErr::new(span, SwErrKind::Int { group, err })
                            }
                            ParseFracErr::NumeratorOverflow { .. } => {
                                unreachable!("max nanosecond has 9 characters, max u32 has 10")
                            }
                        }
                    })?;
                if u64::from(nanos) >= group.max() {
                    unreachable!("max nanosecond has 9 characters. add 1 to max and it has 10 characters. that case is checked previously.");
                }
                dur = dur
                    .checked_add(Duration::from_nanos(nanos.into()))
                    .ok_or_else(|| ParseErr::new(span, SwErrKind::DurationOverflow(group)))?;
            }
        }

        Ok(Self { dur, is_neg })
    }
}

pub(crate) struct SwLexer<'s> {
    content: Peekable<Rev<GraphemeIndices<'s>>>,
    s: &'s str,
}

impl<'s> SwLexer<'s> {
    pub(crate) fn new(s: &'s str) -> Self {
        Self {
            content: UnicodeSegmentation::grapheme_indices(s, true)
                .rev()
                .peekable(),
            s,
        }
    }

    fn advance(content: &mut Peekable<Rev<GraphemeIndices<'s>>>) -> Option<(usize, &'s str)> {
        loop {
            let next = content.next()?;
            if !next.1.chars().all(char::is_whitespace) {
                // only yield non-whitespace grapheme. this prevents prevents
                // Data from having trailing whitespace.
                break Some(next);
            }
        }
    }

    fn peek<'a>(
        content: &'a mut Peekable<Rev<GraphemeIndices<'s>>>,
    ) -> Option<&'a (usize, &'s str)> {
        content.peek()
    }

    fn single_token(next: &str) -> Option<SwTokenKind> {
        match next {
            ":" => Some(SwTokenKind::Colon),
            "." => Some(SwTokenKind::Dot),
            "+" => Some(SwTokenKind::Pos),
            "-" => Some(SwTokenKind::Neg),
            _ => None,
        }
    }
}

impl<'s> Iterator for SwLexer<'s> {
    type Item = SwToken<'s>;

    fn next(&mut self) -> Option<Self::Item> {
        let next = Self::advance(&mut self.content)?;
        let mut span = ByteSpan::new(next.0, next.1.len(), self.s);
        if let Some(typ) = Self::single_token(next.1) {
            Some(SwToken { typ, span })
        } else {
            let mut bytes_ignored = 0;
            while let Some(d_next) = Self::peek(&mut self.content) {
                if Self::single_token(d_next.1).is_some() {
                    break;
                }
                // ignore leading whitespace
                if d_next.1.chars().all(char::is_whitespace) {
                    bytes_ignored += d_next.1.len();
                } else {
                    // oops, not leading whitespace. add all the bytes we ignored to the span.
                    span.shift_start_left(d_next.1.len() + bytes_ignored);
                    bytes_ignored = 0;
                }
                self.content.next();
            }
            Some(SwToken {
                typ: SwTokenKind::Data,
                span,
            })
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum SwTokenKind {
    Colon,
    Dot,
    Pos,
    Neg,
    Data,
}

#[derive(Debug, PartialEq, Eq)]
pub(crate) struct SwToken<'s> {
    pub(crate) typ: SwTokenKind,
    pub(crate) span: ByteSpan<'s>,
}

#[derive(Debug)]
struct Groups<'s>([ByteSpan<'s>; 4]);

impl<'s> Groups<'s> {
    pub(crate) fn new(s: &'s str) -> Self {
        Self([ByteSpan::new(0, 0, s); 4])
    }
}

impl<'s> ops::Index<Group> for Groups<'s> {
    type Output = ByteSpan<'s>;

    fn index(&self, idx: Group) -> &Self::Output {
        &self.0[idx as usize]
    }
}

impl ops::IndexMut<Group> for Groups<'_> {
    fn index_mut(&mut self, idx: Group) -> &mut Self::Output {
        &mut self.0[idx as usize]
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum Group {
    Hours,
    Minutes,
    SecondsInt,
    SecondsSub,
}

impl Group {
    pub(crate) const fn max(self) -> u64 {
        match self {
            Self::Hours => u64::MAX / SEC_PER_HOUR as u64 + 1,
            Self::Minutes => MIN_PER_HOUR as _,
            Self::SecondsInt => SEC_PER_MIN as _,
            #[allow(clippy::cast_possible_truncation)]
            Self::SecondsSub => Duration::from_secs(1).as_nanos() as _,
        }
    }
}

impl fmt::Display for Group {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str(match self {
            Group::Hours => "hours",
            Group::Minutes => "minutes",
            Group::SecondsInt => "seconds",
            Group::SecondsSub => "subseconds",
        })
    }
}
